core/lib/multivm/src/versions/vm_fast/vm.rs

```diff
@@ -503,66 +499,67 @@ impl<S: ReadStorage> VmInterface for Vm<S> {
-                        log: StorageLog {
-                            key: StorageKey::new(AccountTreeId::new(address), u256_to_h256(key)),
-                            value: u256_to_h256(change.after),
-                            kind: if change.is_initial {
-                                StorageLogKind::InitialWrite
-                            } else {
-                                StorageLogKind::RepeatedWrite
-                            },
+        // If the execution is halted, the VM changes are expected to be rolled back by the caller.
+        // Earlier VMs return empty execution logs in this case, so we follow this behavior.
+        let logs = if ignore_world_diff {
+            VmExecutionLogs::default()
+        } else {
+            let storage_logs = self
+                .inner
+                .world_diff
+                .get_storage_changes_after(&start)
+                .map(|((address, key), change)| StorageLogWithPreviousValue {
+                    log: StorageLog {
+                        key: StorageKey::new(AccountTreeId::new(address), u256_to_h256(key)),
+                        value: u256_to_h256(change.after),
+                        kind: if change.is_initial {
+                            StorageLogKind::InitialWrite
+                        } else {
+                            StorageLogKind::RepeatedWrite
                         },
-                        previous_value: u256_to_h256(change.before.unwrap_or_default()),
-                    })
-                    .collect(),
+                    },
+                    previous_value: u256_to_h256(change.before.unwrap_or_default()),
+                })
+                .collect();
+            let events = merge_events(
+                self.inner.world_diff.events_after(&start),
+                self.batch_env.number,
+            );
+            let user_l2_to_l1_logs = extract_l2tol1logs_from_l1_messenger(&events)
+                .into_iter()
+                .map(Into::into)
+                .map(UserL2ToL1Log)
+                .collect();
+            let system_l2_to_l1_logs = self
+                .inner
+                .world_diff
+                .l2_to_l1_logs_after(&start)
+                .iter()
+                .map(|x| x.glue_into())
+                .collect();
+            VmExecutionLogs {
+                storage_logs,
                 events,
                 user_l2_to_l1_logs,
-                system_l2_to_l1_logs: self
-                    .inner
-                    .world_diff
-                    .l2_to_l1_logs_after(&start)
-                    .iter()
-                    .map(|x| x.glue_into())
-                    .collect(),
+                system_l2_to_l1_logs,
                 total_log_queries_count: 0, // This field is unused
-            },
+            }
+        };
+
+        // FIXME: are statistics rolled back on halt as well?
```

Comment:
nitpick: either a tracking issue has to be created and linked here or `FIXME` should be resolved and removed.